src/sympc/tensor/share_control.py

```diff
@@ -0,0 +1,249 @@
+        self.shape = None
+
+        if secret is not None:
+            secret, shape, is_remote_secret = ShareTensorCC.sanity_checks(
+                secret, shape, session
+            )
+            parties = session.parties
+            self.shape = shape
+
+            if is_remote_secret:
+                # If the secret is remote we use PRZS (Pseudo-Random-Zero Shares) and the
+                # party that holds the secret will add it to it's share
+                self.share_ptr = ShareTensorCC.generate_przs(self.shape, self.session)
+                for i, share in enumerate(self.share_ptr):
+                    if share.client == secret.client:
+                        self.share_ptr[i] = self.share_ptr[i] + secret
+                        break
+            else:
+                self.share_ptr = []
+
+                shares = ShareTensorCC.generate_shares(secret, self.session)
+                for share, party in zip(shares, self.session.parties):
+                    self.share_ptr.append(share.send(party))
+
+        elif shares is not None:
+            self.share_ptr = shares
+
+    @staticmethod
+    def sanity_checks(
+        secret: Union[torch.Tensor, float, int],
+        shape: Union[torch.Size, tuple],
+        session: Session,
+    ) -> Tuple[ShareTensor, Union[torch.Size, tuple], bool]:
+        is_remote_secret = False
+
+        if ispointer(secret):
+            is_remote_secret = True
+            if shape is None:
+                raise ValueError(
+                    "Shape must be specified if secret is at another party"
+                )
+        else:
+            if isinstance(secret, (int, float)):
+                secret = torch.tensor(data=[secret])
+
+            if isinstance(secret, torch.Tensor):
+                secret = ShareTensor(data=secret, session=session)
+
+            shape = secret.shape
+
+        return secret, shape, is_remote_secret
+
+    @staticmethod
+    def generate_przs(
+        shape: Union[torch.Size, tuple], session: Session
+    ) -> List[ShareTensor]:
+
+        shape = tuple(shape)
+
+        shares = []
+        for session_ptr, generators_ptr in zip(
+            session.session_ptr, session.przs_generators
+        ):
+            share_ptr = session_ptr.przs_generate_random_share(shape, generators_ptr)
+            shares.append(share_ptr)
+
+        return shares
+
+    @staticmethod
+    def generate_shares(secret, session: Session) -> List[ShareTensor]:
+        if not isinstance(secret, ShareTensor):
+            raise ValueError("Secret should be a ShareTensor")
+
+        parties = session.parties
+        nr_parties = len(parties)
+
+        min_value = session.min_value
+        max_value = session.max_value
+
+        shape = secret.shape
+        tensor_type = session.tensor_type
+
+        random_shares = []
+        for _ in range(nr_parties - 1):
+            rand_value = torch.randint(min_value, max_value, shape, dtype=tensor_type)
```

Comment:
same point about csprng

Reply:
Yep - I opened an issue for it - I thought it would be harder to add the "CSPRNG", but it seems easy peasy - will add it here.